src/modules/ekf2/EKF/pos_vel_estimator_beacon.cpp

```diff
@@ -0,0 +1,168 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "pos_vel_estimator_beacon.hpp"
+
+#include "python/ekf_derivation/generated/compute_beacon_range_fusion.h"
+#include "python/ekf_derivation/generated/predict_covariance_beacon.h"
+
+using namespace time_literals;
+
+void PosVelFromBeacon::init(matrix::Vector<float, NUM_STATES> state_init)
+{
+	if (!_initialised) {
+		_state = state_init;
+		_P.setZero();
+		_P(VX, VX) = MAX_VEL_VARIANCE;
+		_P(VY, VY) = MAX_VEL_VARIANCE;
+		_P(X, X) = MAX_POS_VARIANCE;
+		_P(Y, Y) = MAX_POS_VARIANCE;
+		_initialised = true;
+	}
+}
+void
+PosVelFromBeacon::update(uint64_t time_now, const Vector2f delta_vel, const float dt_prediction)
+{
+	if (!_initialised) {
+		return;
+	}
+
+	predictState(delta_vel, dt_prediction);
+	predictCovariance(time_now, delta_vel);
+}
+
+void PosVelFromBeacon::predictCovariance(uint64_t time_now,
+                                         const Vector2f &delta_vel)
+{// run covariance prediction at a slower rate
+	const float dt_cov_prediction = (float) (time_now - _time_last_update) * 1e-6f;
+	if (time_now - _time_last_update < FILTER_UPDATE_RATE_US || _time_last_update == 0) {
+		if (_time_last_update == 0) {
+			_time_last_update = time_now;
+		}
+
+		return;
+	}
+	_time_last_update = time_now;
+	matrix::Matrix<float, NUM_STATES, NUM_STATES> P_new;
+	const float var = sq(_accel_noise_density * dt_cov_prediction);
+	matrix::Vector2f delta_vel_var(var, var);
+	sym::PredictCovarianceBeacon(_state, _P, delta_vel, delta_vel_var, dt_cov_prediction, &P_new);
+	_P = P_new;
+}
+
+void PosVelFromBeacon::predictState(const Vector2f &delta_vel, const float dt_prediction)
+{// predict state
+	_state(VX, 0) += delta_vel(0);
+	_state(VY, 0) += delta_vel(1);
+	_state(X, 0) += _state(VX, 0) * dt_prediction;
+	_state(Y, 0) += _state(VY, 0) * dt_prediction;
+}
+
+void PosVelFromBeacon::fuseRange(const beaconRangeSample &range_sample, const float vehicle_pos_z)
+{
+	if (!_initialised) {
+		return;
+	}
+
+	matrix::Matrix<float, NUM_STATES, 1> K;
+	matrix::Matrix<float, 1, NUM_STATES> H;
+	matrix::Vector3f beacon_position(range_sample.x, range_sample.y, range_sample.z);
+	const float R_meas = 10.0f;
+	float innovation = 0.0f;
+	float innov_var = 0.0f;
+
+	sym::ComputeBeaconRangeFusion(_state,
+	                              _P,
+	                              vehicle_pos_z,
+	                              beacon_position,
+	                              range_sample.distance,
+	                              R_meas,
+	                              FLT_EPSILON,
+	                              &H,
+	                              &K,
+	                              &innovation,
+	                              &innov_var);
+	_last_innovation = innovation;
+
+	if (innov_var < FLT_EPSILON || check_if_meas_is_rejected(innovation, innov_var, _gate_size)) {
+		return;
+	}
+
+	_state(VX, 0) -= K(VX, 0) * innovation;
```

Comment:
This is a bit awkward, making those vectors will allow you to write `_state += K * innovation`